src/externals/function.rs

```diff
@@ -0,0 +1,186 @@
+            store.inner(),
+            &wasmer::FunctionType::new(argument_types, result_types.clone()),
+            environment,
+            move |environment,
+                  arguments: &[wasmer::Value]|
+                  -> Result<Vec<wasmer::Value>, wasmer::RuntimeError> {
+                let gil = GILGuard::acquire();
+                let py = gil.python();
+
+                let to_py_object = to_py_object(py);
+
+                let arguments: Vec<PyObject> = arguments.iter().map(to_py_object).collect();
+
+                let results = environment
+                    .py_function
+                    .call(py, PyTuple::new(py, arguments), None)
+                    .map_err(|error| {
+                        wasmer::RuntimeError::new(io::Error::from(error).to_string())
+                    })?;
+
+                let result_types = result_types.clone();
+
+                Ok(if let Ok(results) = results.cast_as::<PyTuple>(py) {
+                    results
+                        .iter()
+                        .zip(result_types)
+                        .map(to_wasm_value)
+                        .collect::<PyResult<_>>()
+                        .map_err(|error| {
+                            wasmer::RuntimeError::new(io::Error::from(error).to_string())
+                        })?
+                } else if !results.is_none(py) && result_types.len() > 0 {
+                    vec![to_wasm_value((
+                        results
+                            .cast_as::<PyAny>(py)
+                            .map_err(PyErr::from)
+                            .map_err(|error| {
+                                wasmer::RuntimeError::new(io::Error::from(error).to_string())
+                            })?,
+                        result_types[0],
+                    ))
+                    .map_err(|error| {
+                        wasmer::RuntimeError::new(io::Error::from(error).to_string())
+                    })?]
+                } else {
+                    Vec::new()
+                })
+            },
+        );
+
+        Ok(Self::raw_new(host_function))
+    }
+
+    #[call]
+    #[args(arguments = "*")]
+    fn __call__<'p>(&self, py: Python<'p>, arguments: &PyTuple) -> PyResult<PyObject> {
+        let arguments: Vec<wasmer::Value> = arguments
+            .iter()
+            .zip(self.inner.ty().params().iter().cloned())
+            .map(to_wasm_value)
+            .collect::<PyResult<_>>()?;
+
+        let results = self
+            .inner
+            .call(&arguments)
+            .map(<[_]>::into_vec)
+            .map_err(to_py_err::<RuntimeError, _>)?;
+
+        let to_py_object = to_py_object(py);
+
+        Ok(match results.len() {
+            0 => py.None(),
+            1 => to_py_object(&results[0]),
+            _ => PyTuple::new(
+                py,
+                results.iter().map(to_py_object).collect::<Vec<PyObject>>(),
+            )
+            .to_object(py),
+        })
+    }
```

Comment:
This logic looks like something that could be done with `.collect()` using https://doc.rust-lang.org/std/iter/trait.FromIterator.html ; not required at all, just bringing it up in case you haven't seen that before

Reply:
What part could be replaced by a `.collect()`? I don't get it.

Reply:
The logic of turning results into a PyObject

Reply:
handling 0, 1, or many